Add IQS007 rule (#2370)
home-assistant/developers.home-assistant: docs/core/integration-quality-scale/rules/iqs007.md

---
title: "IQS007: Check during integration initialization if we are able to set it up correctly"
---

IQS007 is as follows:

Check during integration initialization if we are able to set it up correctly

## Reasoning

When we initialize an integration, we should check if we are able to set it up correctly.

> Since not every integration polls for data, for example when they only expose services, we should check so we can let the user know that it doesn't work.

Reviewer: I don't think we should mention an example. Also for polling integrations we want to poll once during setup to check that it works.
```suggestion
This way we can immediately let the user know that it doesn't work.
```

Implementing these checks increases the confidence that the integration will work correctly and provides a user-friendly way to show errors.

> This will increase the user experience.

Reviewer: ```suggestion
This will improve the user experience.
```

## Example implementation

> When the reason of the failure is temporary (like a temporary offline device), we should raise `ConfigEntryNotReady` and Home Assistant will retry the setup later.

Reviewer: ```suggestion
When the reason for the failure is temporary (like a temporary offline device), we should raise `ConfigEntryNotReady` and Home Assistant will retry the setup later.
```

> If the reason of the failure is that the password is incorrect or the api key is invalid, we should raise `ConfigEntryAuthFailed` and Home Assistant will ask the user to reauthenticate (if they have the reauthentication flow implemented).

Reviewer: ```suggestion
If the reason for the failure is that the password is incorrect or the api key is invalid, we should raise `ConfigEntryAuthFailed` and Home Assistant will ask the user to reauthenticate (if the reauthentication flow is implemented).
```

If we don't expect the integration to work in the foreseeable future, we should raise `ConfigEntryError`.

`__init__.py`:
```python
async def async_setup_entry(hass: HomeAssistant, entry: MyIntegrationConfigEntry) -> bool:
    """Set up my integration from a config entry."""

    client = MyClient(entry.data[CONF_HOST])
    
    try:
        await client.async_setup()
    except OfflineException:
        raise ConfigEntryNotReady("Device is offline")
    except InvalidAuthException:
        raise ConfigEntryAuthFailed("Invalid authentication")
    except AccountClosedException:
        raise ConfigEntryError("Account closed")

    entry.runtime_data = coordinator

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True
```

## Additional resources

More information about config entries and their lifecycle can be found in the [config entry documentation](../../../config_entries_index).

## Exceptions

> There are no exceptions to this rule.

Author: Same concern as https://github.com/home-assistant/developers.home-assistant/pull/2368/files#r1799719451
